src/constraint.rs

```diff
@@ -703,3 +715,65 @@ impl ConstraintValidator for FieldsConstraint {
         Ok(())
     }
 }
+
+/// Implements Ion Schema's `contains` constraint
+/// [contains]: https://amzn.github.io/ion-schema/docs/spec.html#contains
+#[derive(Debug, Clone, PartialEq)]
+pub struct ContainsConstraint {
+    // TODO: convert this into a HashSet once we have an implementation of Hash for OwnedElement in ion-rust
+    // Reference ion-rust issue: https://github.com/amzn/ion-rust/issues/220
+    values: Vec<OwnedElement>,
+}
+
+impl ContainsConstraint {
+    pub fn new(values: Vec<OwnedElement>) -> Self {
+        Self { values }
+    }
+}
+
+impl ConstraintValidator for ContainsConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // get the expected values from the contains constraint
+        let mut expected_values = self.values.to_owned();
+
+        // Check for null sequence
+        if value.is_null() {
+            return Err(Violation::new(
+                "contains",
+                ViolationCode::TypeMismatched,
+                &format!("expected a sequence found {:?}", value),
+            ));
+        }
+
+        match value.as_sequence() {
+            None => {
+                // return Violation if value is not an Ion sequence
+                return Err(Violation::new(
+                    "contains",
+                    ViolationCode::TypeMismatched,
+                    &format!("expected list/sexp found {}", value.ion_type()),
+                ));
+            }
+            Some(ion_sequence) => {
+                // for each value in ion sequence if it exists in the expected values
+                // then remove it from expected values to keep track of missing values
+                for value in ion_sequence.iter() {
+                    if let Some(pos) = expected_values.iter().position(|v| v == value) {
+                        expected_values.remove(pos);
+                    }
+                }
+
+                // return Violation if there were any values left in the expected values vector
+                if !expected_values.is_empty() {
+                    return Err(Violation::new(
+                        "contains",
+                        ViolationCode::MissingValue,
+                        &format!("{:?} has missing value(s): {:?}", value, expected_values),
+                    ));
+                }
```

Comment:
You can eliminate the need to own `expected_values` if you use a new local vec for storing missing values (rather than removing expected values that are not found).
```suggestion
                // for each value in ion sequence if it does not exist in the expected values
                // then add it to missing_values to keep track of missing values
                let mut missing_values = Vec::new();
                for expected_value in expected_values.iter() {
                    if None == ion_sequence.iter().position(|v| v == expected_value) {
                        missing_values.push(expected_value);
                    }
                }

                // return Violation if there were any values added to the missing values vector
                if !missing_values.is_empty() {
                    return Err(Violation::new(
                        "contains",
                        ViolationCode::MissingValue,
                        &format!("{:?} has missing value(s): {:?}", value, missing_values),
                    ));
                }
```